apps/examples/src/examples/sync-private-content/SyncPrivateContentExample.tsx

```diff
@@ -0,0 +1,146 @@
+import { useSyncDemo } from '@tldraw/sync'
+import React from 'react'
+import {
+	Atom,
+	TLComponents,
+	Tldraw,
+	react,
+	useAtom,
+	useEditor,
+	useIsToolSelected,
+	useTools,
+	useValue,
+} from 'tldraw'
+import 'tldraw/tldraw.css'
+import { VisibilityOff, VisibilityOn } from '../../icons/icons'
+import { Toggle } from './Toggle'
+import './style.css'
+
+// There's a guide at the bottom of this file!
+
+// [1]
+const PrivateModeContext = React.createContext<null | Atom<boolean>>(null)
+
+const components: TLComponents = {
+	// [2]
+	InFrontOfTheCanvas: () => {
+		const editor = useEditor()
+		const isInSelectTool = useIsToolSelected(useTools().select)
+		const userId = useValue('userId', () => editor.user.getId(), [])
+		const myPrivateSelectedShapes = useValue(
+			'private shapes',
+			() =>
+				editor
+					.getSelectedShapes()
+					.filter((shape) => !!shape.meta.private && shape.meta.ownerId === userId),
+			[editor, userId]
+		)
+
+		// [3]
+		const isPrivateDrawingMode$ = React.useContext(PrivateModeContext)!
+		const isPrivateDrawingMode = useValue(isPrivateDrawingMode$)
+
+		return (
+			<>
+				{isInSelectTool && myPrivateSelectedShapes.length > 0 ? (
+					<div className="toggle-panel">
+						<div>
+							Make {myPrivateSelectedShapes.length} selected shape
+							{myPrivateSelectedShapes.length > 1 ? 's' : ''} public?{' '}
+						</div>
+						<button
+							onClick={() => {
+								editor.markHistoryStoppingPoint()
+								// [7]
+								const allAffectedShapes = [
+									...editor.getShapeAndDescendantIds(myPrivateSelectedShapes.map((s) => s.id)),
+								].map((id) => editor.getShape(id)!)
+								editor.updateShapes(
+									allAffectedShapes.map((shape) => ({
+										...shape,
+										meta: { ...shape.meta, private: false },
+									}))
+								)
+							}}
+						>
+							Yes
+						</button>
+					</div>
+				) : (
+					<div
+						className="toggle-panel pointer"
+						onClick={() => isPrivateDrawingMode$.update((v) => !v)}
+					>
+						{isPrivateDrawingMode ? <VisibilityOff fill="#444" /> : <VisibilityOn fill="#444" />}
+						<div>Private mode</div>
+						<Toggle isChecked={isPrivateDrawingMode} />
+					</div>
+				)}
+			</>
+		)
+	},
+}
+function App({ roomId }: { roomId: string }) {
+	const store = useSyncDemo({ roomId })
+	const isPrivateDrawingMode$ = React.useContext(PrivateModeContext)!
+	return (
+		<div className="tldraw__editor">
+			<Tldraw
+				store={store}
+				deepLinks
+				// [4]
+				isShapeHidden={(shape, editor) => {
+					const userId = editor.user.getId()
+					return !!shape.meta.private && shape.meta.ownerId !== userId
+				}}
+				onMount={(editor) => {
+					// [5]
+					editor.store.sideEffects.registerBeforeCreateHandler('shape', (shape) => {
+						if ('private' in shape.meta) return shape
+						return {
+							...shape,
+							meta: {
+								...shape.meta,
```

Comment:
same q re: spreading `shape.meta` here

Reply:
here is different. we don't do any post-processing on the return values of store side effects. it's a much lower-level API. so you need to do a spread here.

Reply:
ah good to know!